src/conv/heuristics/ai_heuristics.cpp

```
@@ -0,0 +1,445 @@
+{
+    std::vector<V> values = {};
+    values.reserve(keys.size());
+    std::transform(keys.begin(), keys.end(), std::back_inserter(values), [&](const U& key) {
+        return map.at(key);
+    });
+    return values;
+}
+} // namespace common
+
+#if MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK
+namespace immed_mode {
+Metadata::Metadata(const std::string& arch)
+    : json(common::LoadJSON(GetSystemDbPath() + "/" + arch + "_metadata.tn.model")),
+      direction_encodings(json["encodings"]["Direction"]),
+      precision_encodings(json["encodings"]["Precision"]),
+      layout_encodings(json["encodings"]["Layout"]),
+      features(json["conv_params_used_as_features"]),
+      num_inputs(json["num_inputs"]),
+      num_outputs(json["num_outputs"]),
+      num_solvers(json["num_solvers"]),
+      solver_map(common::ReverseMap<std::string, size_t>(json["encodings"]["solver"])),
+      features_mean(common::LookupValues<std::string, float>(
+          features, json["stats"]["overall"]["features"]["mean"])),
+      features_std(common::LookupValues<std::string, float>(
+          features, json["stats"]["overall"]["features"]["std"]))
+{
+}
+
+size_t Metadata::EncodeDirection(const miopen::conv::Direction& dir) const
+{
+    if(dir == conv::Direction::BackwardWeights)
+        return direction_encodings.at("W");
+    else if(dir == conv::Direction::BackwardData)
+        return direction_encodings.at("B");
+    else
+        return direction_encodings.at("F");
+}
+size_t Metadata::EncodePrecision(const miopenDataType_t& data_type) const
+{
+    if(data_type == miopenBFloat16)
+        return precision_encodings.at("BF16");
+    else if(data_type == miopenHalf)
+        return precision_encodings.at("FP16");
+    else
+        return precision_encodings.at("FP32");
+}
+size_t Metadata::EncodeLayout(const std::string& layout) const
+{
+    if(layout == "NCDHW")
+        return layout_encodings.at("NCDHW");
+    else
+        return layout_encodings.at("NCHW");
```

If the layout is neither `NCHW` nor `NCDHW`, it will be considered `NCHW`. Is that how it's intended?

---

If it's neither NCHW or NCDHW TunaNet will never run. See the IsProblemSupported() within this same file.

---

Other layouts can be added later. It is better to know when something goes wrong than to silently ignore the error. Let's add `MIOPEN_THROW(miopenStatusInternalError)` for other layouts.

---

```suggestion
size_t Metadata::EncodeLayout(const std::string& layout) const
{
    if(layout != "NCHW" && layout != 'NCDHW')
        MIOPEN_THROW(miopenStatusInternalError);
    return layout_encodings.at(layout);
```